src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
  * @typedef {import('@web3-storage/gateway-lib').UnixfsContext} UnixfsContext
  */
 
+/**
+ *
+ * @param {string} s
+ * @returns {import('./bindings.js').TokenMetadata}
+ */
+function deserializeTokenMetadata (s) {
+  // TODO should this be dag-json?
+  return JSON.parse(s)
+}
+
+/**
+ *
+ * @param {import('./bindings.js').TokenMetadata} m
+ * @returns string
+ */
+function serializeTokenMetadata (m) {
+  // TODO should this be dag-json?
+  return JSON.stringify(m)
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} cid
+ * @returns {Promise<import('./constants.js').RATE_LIMIT_EXCEEDED>}
+ */
+async function checkRateLimitForCID (env, cid) {
+  const rateLimiter = env.RATE_LIMITER
+  if (!rateLimiter) {
+    console.warn('no rate limiter found')
+    return RATE_LIMIT_EXCEEDED.NO
+  }
+  const rateLimitResponse = await rateLimiter.limit({ key: cid.toString() })
+  if (rateLimitResponse.success) {
+    return RATE_LIMIT_EXCEEDED.NO
+  } else {
+    console.log(`limiting CID ${cid}`)
+    return RATE_LIMIT_EXCEEDED.YES
+  }
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {string} authToken
+ * @returns TokenMetadata
+ */
+async function getTokenMetadata (env, authToken) {
+  const cachedValue = await env.AUTH_TOKEN_METADATA.get(authToken)
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
```

Comment:
**suggestion:** Let's not leave this as a `TODO` here; let's capture the user value and move it to a card we can prioritize on the board. We can't decide to actually do this until it's in the flow and context of our other work.